### PR TITLE
SPARK-4267 [CORE] Failing to launch jobs on Spark on YARN with Hadoop 2.5.0 or later

### PR DESCRIPTION
**Pull request by srowen**
Per SPARK-4267, I think the problem here is stripping quotes in JVM arguments before sending them on the YARN et al. The proposed fix is straightforward.

I'm seeing this make `MQTTStreamSuite` fail though, but can't for the life of me see how it's related. I want to see what Jenkins thinks.

### CODE REVIEW

**Comment by andrewor14** [comment]
Hey @srowen have you tested it before and after this patch to see if it actually fixes it? If this is a problem in master it's probably also a problem in older branches, since the `splitCommandString` method hasn't changed that much.

Also from the JIRA it seems that someone also ran into something similar for 2.4.0. Does this happen only since Hadoop 2.5.0? Though IIRC I used to run branch-1.1 successfully with 2.4.0 when I was doing the release for 1.1.1.

**Comment by srowen** [comment]
@andrewor14 I reproduced this on Spark 1.2.0 and verified that it was the `-Dnumbers="one two three"` arg that triggers the problem. The executors fail to start. I haven't rerun on my cluster with the fix though, no. I suspect it's a problem for lots of versions, but only if you specify arguments like `-Dnumbers="one two three"`

**Comment by andrewor14** [comment]
I see. That's not surprising considering how YARN runs commands as strings rather than sequence of strings, from what I've heard at least.

**Comment by srowen** [comment]
If you mean that YARN doesn't parse the argument like bash does, I think that's true. But the issue here is that the quotes have disappeared by the time the command line args leave Spark, so there's no chance that `-Dnumbers="one two three"` is parsed as one arg, since it's `-Dnumbers=one two three` after it's split and rejoined.

**Comment by andrewor14** [comment]
Yeah, meaning the quotes in `-Dnumbers="one two three"` are not escaped properly and the setting is not treated as one argument.

**Comment by srowen** [comment]
PS @andrewor14 I did run my patched version on a cluster, with and without the offending argument, using a command like that reported in SPARK-4267, and it succeeded both times. Reasonable evidence this is the fix. Yes I suppose you could consider re-quoting any parsed argument that has a space. I thought it might be less surprising and error-prone to just not modify the command string, so that simply joining on ' ' then gives you back the original command (well, maybe without a run of excess whitespace in between args).

**Comment by andrewor14** [comment]
Ok thanks for verifying. I will take a closer look later today

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -1283,6 +1291,7 @@ private[spark] object Utils extends Logging {
     if (inWord || inDoubleQuote || inSingleQuote) {
       endWord()
     }
+    println("+++ split command to " + buf)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: probably shouldn't print this

Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Oof, sorry, can't believe I left that in!

**Comment by andrewor14** [comment]
Hey @srowen it seems like this will break existing behavior though. What if I want to run an application with the following arguments
```
a "b c" d
```
and I want them to be parsed as `a`, `b c`, and `d` without the quotes? I don't see a way to do that now but maybe I'm missing something

**Comment by andrewor14** [comment]
To fix the YARN issue maybe we should do something specific there, like escaping the double quotes before passing them to YARN?

**Comment by srowen** [comment]
@andrewor14 Yeah basically this keeps the quotes all the way through. It's not a problem of escaping, but the opposite -- the quotes are gone from the get-go. Yes, that's an issue and probably more of a risk than I figured.

Another way to go is try to re-quote strings with spaces when they're re-joined. I suppose that `-Dfoo="bar baz"` becomes `"-Dfoo=bar baz"`, but AFAIK this will parse the same way in bash. What about that kind of solution?

**Comment by vanzin** [comment]
Hi @srowen,

I think this isn't the correct fix. I think the proper fix would be to change this in `Client.scala`:

    sparkConf.getOption("spark.driver.extraJavaOptions")
      .orElse(sys.env.get("SPARK_JAVA_OPTS"))
      .map(Utils.splitCommandString).getOrElse(Seq.empty)
      .foreach(opts => javaOpts += opts)

Into this:

    val opts sparkConf.getOption("spark.driver.extraJavaOptions")
      .orElse(sys.env.get("SPARK_JAVA_OPTS"))
      .map(Utils.splitCommandString)
      .foreach(opt => javaOpts += YarnSparkHadoopUtil.escapeForShell(opt))

`ExecutorRunnable.scala` has similar code, but it just treats the command line as a string instead of a list, so it doesn't need to parse the options into a list then escape them back when building the command line.

`Utils.splitCommandString` is currently doing what it's supposed to do; parsing a command line into a list of individual arguments that can be fed to `Runtime.exec()`, for example. So it has to strip quotes that the shell would interpret, otherwise you break code that does that (e.g. I believe standalone mode uses `ProcessBuilder` to run executors).

`YarnSparkHadoopUtil.escapeForShell` escapes individual arguments back into something that a shell would interpret as a single argument. This is needed since Yarn launches containers by creating a shell script to do the work, instead of using `ProcessBuilder`.

**Comment by vanzin** [comment]
BTW I'd also recommend adding a test to `YarnClusterSuite` (or extending an existing one).

**Comment by vanzin** [comment]
BTW#2, the code that handles `spark.yarn.am.extraJavaOptions` a few lines below the code I mentioned has the same issue.

**Comment by srowen** [comment]
Yeah I don't think this is the way to do it. Scratch this one; let me have another run at this problem from the other end.

**Comment by srowen** [comment]
@vanzin the code in `ExecutorRunnable.scala` that handles the executor opts still does build up `javaOpts` as a collection.  You're saying it doesn't need this treatment because that `List[String]` goes straight to `setCommands` in the YARN API? The thing is, it is exactly `spark.executor.extraJavaOpts` that was the problem, so I imagine something has to change. You made an additional comment that YARN actually won't escape these arguments, so we should? this makes me dizzy.

**Comment by vanzin** [comment]
Yeah, Yarn makes a mess out of this. Basically, it does this:

    val args = List("abd", "def", "g h")
    val cmd = args.mkString(" ")

So even though you have 3 arguments in `args`, the actual execution will have 4 when interpreted by the shell. So the code in `ExecutorRunnable` works because it just builds a big command line, instead of trying to parse that command line into a proper argument list with `Utils.splitCommandString`.

Anyway, Yarn's API is broken but it is what it is, so we have to work with it. It would be nice if both `Client` and `ExecutorRunnable` used the same approach, for consistency, but not needed to fix this issue.

**Comment by srowen** [comment]
@vanzin I don't think it does build a big command line though. I see:

```
    val commands = prepareCommand(...) // List[String]
    ...
    ctx.setCommands(commands)
```

The elements of that list can and should be escaped?
If I update `YarnClusterSuite` to add a `spark.executor.extraJavaOptions` like `-Dfoo="one two three"` then the test fails... although just adding this escaping routine to the args in `ExecutorRunnable` too does not fix that. Still investigating...

**Comment by vanzin** [comment]
I was basing my comments on this code:

    // Set extra Java options for the executor, if defined
    sys.props.get("spark.executor.extraJavaOptions").foreach { opts =>
      javaOpts += opts
    }

That code isn't "unescaping" the Spark configuration, so whatever you set in `spark.executor.extraJavaOptions` should, theoretically, show up verbatim on the executor's command line. But I haven't run any tests, so if tests who otherwise, please do change it.

**Comment by srowen** [comment]
@vanzin Got it, yes. Yes, this has to change too, to parse the arguments and then escape them, or else escaping them does the wrong thing, and things have to be escaped to solve the original problem here. I'm going to fix up several bits of code around here to treat these args / parsing / escaping consistently. good news is it makes the tests, including one that exercises this problem, pass. PR coming...

**Pull request closed without merging**